src/bootstrap.rs

```diff
@@ -30,3 +30,64 @@ pub fn start() {
         let _ = connect::connect_to(proxy, None, Some(&maker));
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::test_utils::new_random_qp2p_for_unit_test;
+    use crate::{Builder, Config, Event};
+    use std::collections::HashSet;
+    use std::net::{IpAddr, Ipv4Addr};
+    use std::sync::mpsc;
+    use std::thread;
+    use std::time::Duration;
+
+    // Try to bootstrap to 4 different peer at once. Use an artificial delay for 3 of the peers.
+    // Make sure in the end we have only one bootstrapped connection (all other connections are
+    // supposed to be dropped).
+    #[test]
+    fn bootstrap_to_only_one_node() {
+        const NUM_CONNS: u64 = 4;
+        const DELAY_MS: u64 = 10;
+
+        let mut bs_nodes = Vec::new();
+        let mut bs_peer_addrs = Vec::new();
+        let mut hcc_contacts = HashSet::new();
+
+        // Spin up 4 peers that we'll use for hardcoded contacts.
+        for i in 0..NUM_CONNS {
+            let (tx, rx) = mpsc::channel();
+            let mut qp2p = unwrap!(Builder::new(tx)
+                .with_config(Config {
+                    ip: Some(IpAddr::V4(Ipv4Addr::LOCALHOST)),
+                    port: Some(0),
+                    ..Default::default()
+                })
+                .build());
+
+            qp2p.set_connect_delay(i * DELAY_MS);
```

Comment:
In fact just to be robust, i would say have a few connect without delay too (so simultaneous) and make sure that we still get one connection. You can make this a private function taking delay as optional parameter. If None then set it to zero else use the supplied one - that way you will have two test cases - one this exact one simulated and the other without delay.

If the Delay is 0 then ofc sleep for some non-zero interval in the code below